stan/math/prim/fun/multiply.hpp

```diff
@@ -15,71 +16,90 @@ namespace math {
  * @return product of matrix and scalar
  */
-template <int R, int C, typename T1, typename T2,
-          typename = require_all_arithmetic_t<T1, T2>>
-inline Eigen::Matrix<return_type_t<T1, T2>, R, C> multiply(
-    const Eigen::Matrix<T1, R, C>& m, T2 c) {
+template <typename T1, typename T2, typename = require_eigen_t<T1>,
+          typename = require_stan_scalar_t<T2>>
+inline auto multiply(const T1& m, T2 c) {
   return c * m;
 }
 
 /**
  * Return specified scalar multiplied by specified matrix.
  *
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
- * @tparam T1 type of scalar
- * @tparam T2 type of elements in the matrix
+ * @tparam T1 type of the scalar
+ * @tparam T2 type of the matrix or expression
  *
  * @param c scalar
  * @param m matrix
  * @return product of scalar and matrix
  */
-template <int R, int C, typename T1, typename T2,
-          typename = require_all_arithmetic_t<T1, T2>>
-inline Eigen::Matrix<return_type_t<T1, T2>, R, C> multiply(
-    T1 c, const Eigen::Matrix<T2, R, C>& m) {
+template <typename T1, typename T2, typename = require_stan_scalar_t<T1>,
+          typename = require_eigen_t<T2>>
+inline auto multiply(T1 c, const T2& m) {
   return c * m;
 }
 
 /**
- * Return the product of the specified matrices.  The number of
+ * Return the product of the specified matrices. The number of
  * columns in the first matrix must be the same as the number of rows
  * in the second matrix.
  *
- * @tparam R1 number of rows in the first matrix, can be Eigen::Dynamic
- * @tparam C1 number of columns in the first matrix, can be Eigen::Dynamic
- * @tparam R2 number of rows in the second matrix, can be Eigen::Dynamic
- * @tparam C2 number of columns in the second matrix, can be Eigen::Dynamic
- * @tparam T1 type of elements in first matrix
- * @tparam T2 type of elements in second matrix
+ * @tparam T1 type of the first matrix or expression
+ * @tparam T2 type of the second matrix or expression
  *
- * @param m1 first matrix
- * @param m2 second matrix
+ * @param m1 first matrix or expression
+ * @param m2 second matrix or expression
  * @return the product of the first and second matrices
  * @throw <code>std::invalid_argument</code> if the number of columns of m1 does
  * not match the number of rows of m2.
  */
-template <int R1, int C1, int R2, int C2, typename T1, typename T2,
-          typename = require_all_arithmetic_t<T1, T2>>
-inline Eigen::Matrix<return_type_t<T1, T2>, R1, C2> multiply(
-    const Eigen::Matrix<T1, R1, C1>& m1, const Eigen::Matrix<T2, R2, C2>& m2) {
+template <typename T1, typename T2, typename = require_all_eigen_t<T1, T2>,
+          typename
+          = require_all_arithmetic_t<value_type_t<T1>, value_type_t<T2>>,
+          typename
+          = require_any_not_same_t<double, value_type_t<T1>, value_type_t<T2>>,
+          typename = require_not_t<
+              conjunction<is_eigen_row_vector<T1>, is_eigen_col_vector<T2>>>>
```

Comment:
I'm not totally understanding the concept on this one

```
template <typename Mat1, typename Mat2, typename = require_all_eigen_vt<std::is_arithmetic, Mat1, Mat2>
```

Then specialize for one being an integral? (since the above previously said this is for a subset of arithmetic)

Let's add that `is_dot_product` to clean these up

Reply:
We can't use specialization, since return type is different (this one returns expression, all doubles returns a matrix). So it must be an overload.